Update backward definition for more operators and reenable tests in test_ops.py

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44744 [WIP] Add tests for R to C cases
* **#44444 Update backward definition for more operators and reenable tests in test_ops.py**

This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow` (fixes #41510 ), `reciprocal`, `angle`.
2. Re-enables the tests in `test_ops.py`.
3. Adds dispatch for complex dtypes for `tanh_backward`.
4. Re-enables commented tests in `common_methods_invocation.py`.

Differential Revision: [D24027761](https://our.internmc.facebook.com/intern/diff/D24027761)

anjali411 committed:
Update backward definition for more operators and reenable tests in test_ops.py

[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`.
2. Re-enables the tests in `test_ops.py`.
3. Fixes a bug in `.pow()` for the case when exp is complex and equals 1.0 https://github.com/pytorch/pytorch/pull/44444/files#diff-9fa1985487308c7241bba078a8622c44L33 .
4. Adds dispatch for complex dtypes for `tanh`.
5. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`.
2. Re-enables the tests in `test_ops.py`.
3. Fixes a bug in `.pow()` for the case when exp is complex and equals 1.0 https://github.com/pytorch/pytorch/pull/44444/files#diff-9fa1985487308c7241bba078a8622c44L33 .
4. Adds dispatch for complex dtypes for `tanh`.
5. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`.
2. Re-enables the tests in `test_ops.py`.
3. Fixes a bug in `.pow()` for the case when exp is complex and equals 1.0 https://github.com/pytorch/pytorch/pull/44444/files#diff-9fa1985487308c7241bba078a8622c44L33 .
4. Adds dispatch for complex dtypes for `tanh_backward`.
5. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`.
2. Re-enables the tests in `test_ops.py`.
3. Fixes a bug in `.pow()` for the case when exp is complex and equals 1.0 https://github.com/pytorch/pytorch/pull/44444/files#diff-9fa1985487308c7241bba078a8622c44L33 .
4. Adds dispatch for complex dtypes for `tanh_backward`.
5. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`.
2. Re-enables the tests in `test_ops.py`.
3. Fixes a bug in `.pow()` for the case when exp is complex and equals 1.0 https://github.com/pytorch/pytorch/pull/44444/files#diff-9fa1985487308c7241bba078a8622c44L33 .
4. Adds dispatch for complex dtypes for `tanh_backward`.
5. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`, `angle`.
2. Re-enables the tests in `test_ops.py`.
3. Fixes a bug in `.pow()` for the case when exp is complex and equals 1.0 https://github.com/pytorch/pytorch/pull/44444/files#diff-9fa1985487308c7241bba078a8622c44L33 .
4. Adds dispatch for complex dtypes for `tanh_backward`.
5. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`, `angle`.
2. Re-enables the tests in `test_ops.py`.
3. Adds dispatch for complex dtypes for `tanh_backward`.
4. Re-enables commented tests in `common_methods_invocation.py`.


[ghstack-poisoned]

albanD commented on aten/src/ATen/native/Pow.cpp:
```diff
@@ -27,11 +27,8 @@ Tensor& pow_out(Tensor& result, const Tensor& base, Scalar exp) {
+  auto exponent = (exp.isComplex()) ? exp.toComplexDouble() : exp.toDouble();
+  if (exponent == 0.0) {
     result.resize_as_(base).fill_(1);
   } else if (!exp.isComplex() && (exp.toDouble() == 1.0)) {
```
This should be updated as well no? otherwise the case for complex == 1 is not handles anymore.

albanD commented on aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:
```diff
@@ -589,7 +589,20 @@ void logit_backward_kernel(TensorIterator& iter, Scalar eps_scalar) {
 void tanh_backward_kernel(TensorIterator& iter) {
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(iter.dtype(), "tanh_backward_cpu", [&]() {
+  if (isComplexType(iter.dtype())) {
+    AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "tanh_backward_cpu", [&]() {
```
Why is this one separated? Isn't the conj a no-op for reals? Or is the problem that it returns a complex?

anjali411 replied:
`std::conj` returns a complex number for real numbers as well https://en.cppreference.com/w/cpp/numeric/complex/conj. so because it's not a no op for real numbers, I split up dispatch for complex and non-complex dtypes.

albanD commented on test/test_ops.py:
```diff
@@ -90,19 +90,19 @@ def _gradgrad_test_helper(self, device, dtype, op, variant):
     # Tests that gradients are computed correctly
     # TODO(@anjali411) enable this for torch.cdouble.
-    @dtypes(torch.double)
+    @dtypes(torch.double, torch.cdouble)
```
remove the TODO

albanD commented on test/test_autograd.py:
```diff
@@ -4795,7 +4795,8 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
 # and only run for floating point
 
 # TODO(@anjali411): add the commented tests back after updating the formula based on tensorflow definition
```
remove this as well.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1661,8 +1661,8 @@
-  grad_output: tanh_backward(grad, output)
-  output: -2 * output * grad * grad_output
+  grad_output: tanh_backward(grad, output.conj())
+  output: grad.conj() * (-2 * output.conj() * grad_output)
```
Is the re-ordering important here?

anjali411 replied:
no I just re-ordered it because I think it's nice to follow a general convention of `grad * ...` whenever possible for better code readability.

albanD replied:
It looks ok here. We just need to be careful as some of the orderings are important for numerical precision. But this one should be fine.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -198,21 +208,49 @@ Tensor pow_backward_self(Tensor grad, const Tensor & self, const Tensor & expone
 // Currently, tensorflow agrees with us.
 Tensor pow_backward_exponent(Tensor grad, const Tensor& self, const Tensor& exponent, Tensor result) {
-  return grad * at::where(at::logical_and(self == 0, exponent >= 0),
+  auto cond = [](auto self, auto exp) {
```
Why did you put this in a lambda?
Same in the function below.

anjali411 replied:
The condition input to `at::where` is different for complex and non-complex inputs. I thought it's better to put it in a lambda function as opposed to global scope since it's small and only used once.

albanD replied:
But this lambda is never re-used and just called once one line below. So why not just remove it just use the code inside directly and pass the computed Tensor directly to the next line?

albanD replied:
But this lambda is only ever called once on the next line. So why not just leave the content without the lambda and use the computed Tensor on the next line?

anjali411 replied:
makes sense will do

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -198,21 +208,49 @@ Tensor pow_backward_self(Tensor grad, const Tensor & self, const Tensor & expone
 }
 
+Tensor angle_backward(Tensor grad, const Tensor& self) {
+  auto out = grad * self / self.abs().pow(2);
```
This looks different that the previous formula (casting of the gradient in particular). Is that expected?

anjali411 replied:
The previous formula didn't work correctly with real numbers. Actually on more thought, `torch.angle(real_tensors)` always returns zeros, so we can simply return grad (as can also be seen when the formula is simplified).

albanD replied:
Also this will return nan at 0. Is that expected?

albanD replied:
Also this will return nan at 0. Is that expected?

anjali411 replied:
good catch. No, I don't think it's correct. It should return zero for the case when `abs = 0`. updated!

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow`, `reciprocal`, `angle`.
2. Re-enables the tests in `test_ops.py`.
3. Adds dispatch for complex dtypes for `tanh_backward`.
4. Re-enables commented tests in `common_methods_invocation.py`.

Differential Revision: [D24027761](https://our.internmc.facebook.com/intern/diff/D24027761)

[ghstack-poisoned]

anjali411 committed:
Update on "Update backward definition for more operators and reenable tests in test_ops.py"


This PR:
1. Fixes https://github.com/pytorch/pytorch/issues/41510. Updates backward formula for the following functions: `asin`, `acos`, `asinh`, `acosh`, `atan`, `atanh`, `div`, `log`, `log10`, `log2`, `log1p`, `pow` (fixes #41510 ), `reciprocal`, `angle`.
2. Re-enables the tests in `test_ops.py`.
3. Adds dispatch for complex dtypes for `tanh_backward`.
4. Re-enables commented tests in `common_methods_invocation.py`.

Differential Revision: [D24027761](https://our.internmc.facebook.com/intern/diff/D24027761)

[ghstack-poisoned]

albanD reviewed (approved):
lgtm

ezyang commented:
Unfortunately this is failing Windows only test:

```
======================================================================
FAIL [0.017s]: test_grad_scaling_penalty (__main__.TestCuda)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_cuda.py", line 2145, in test_grad_scaling_penalty
    self._run_scaling_case(run, unskipped=3, skipped=1)
  File "test_cuda.py", line 2036, in _run_scaling_case
    self.assertTrue(torch.allclose(c, s, atol=atol))
AssertionError: False is not true

----------------------------------------------------------------------
```

anjali411 commented:
> Unfortunately this is failing Windows only test:
> 
> ```
> ======================================================================
> FAIL [0.017s]: test_grad_scaling_penalty (__main__.TestCuda)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "test_cuda.py", line 2145, in test_grad_scaling_penalty
>     self._run_scaling_case(run, unskipped=3, skipped=1)
>   File "test_cuda.py", line 2036, in _run_scaling_case
>     self.assertTrue(torch.allclose(c, s, atol=atol))
> AssertionError: False is not true
> 
> ----------------------------------------------------------------------
> ```

I think it was a flaky failure or caused by another PR :( because #46275 doesn't have any windows failure.
